### PR TITLE
Add support for Schnorr crypto (BCH variant)

### PR DESCRIPTION
**Pull request by Fabcien**
Support Schnorr signature and verification defined in
https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/2019-05-15-schnorr.md.

The implementation is based on https://github.com/Bitcoin-ABC/secp256k1
as well as the test vectors.

This is a first step toward adding support for Schnorr signature and verification to trezorlib.

### CODE REVIEW

**Comment by onvej-sl** [comment]
Are you able to tell the difference between these signatures and Schnorr signatures defined in https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki, @Fabcien ?

**Comment by Fabcien** [comment]
The goal is the same, but it differs in some choices being made:
 - The computation uses the 33 bytes compressed public key format (vs 32 bytes X only for BIP340)
 - The R.y is chosen as the one being a quadratic residue (vs the one being even for BIP340)

**Comment by prusnak** [comment]
> The goal is the same, but it differs in some choices being made:

Would it be possible to rewrite your code in such a way it supports both BCH Schnorr and BIP-340 Schnorr? For example by adding `schnorr_type` enum with values `SCHNORR_BCH` and `SCHNORR_BIP340`

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:220]: Is (r != 0 and s != 0) actually needed? It's not written in the specification.

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:34]: Please initialise _all_ local variables. It's our convention.

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:134]: `sizeof(Rxbuf)` is better than 32

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:54]: Why don't you use `bn_copy` instead of `bn_cmov`?

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:47]: You should comment a bit what is happening here. Something like:
jacobi(2 * n, k) = jacobi(n, k) if k == 1 (mod 8) or k == 7 (mod 8)
jacobi(2 * n, k) = -jacobi(n, k) if k == 3 (mod 8) or k == 5 (mod 8)

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:56]:
```
jacobi(n, k) = jacobi(k, n) if k = n = 1 (mod 8)
jacobi(n, k) = -jacobi(k, n) if k = n = 3 (mod 8)
```

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:100]: `SHA256_DIGEST_LENGTH` is better than 32.

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:173]: I think `k` is already normalized. Maybe you wanted to reduce it?

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Removed in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c:47]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c:60]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in bc6267c

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Neither is necessary, k is guaranteed to be < order. Removed in bc6267c

**Comment by Fabcien** [comment]
@prusnak It can probably be done, but there are enough differences that it might be cleaner to just create new functions for BIP340. Code factorization can happen at this time to avoid duplication as needed.

**Comment by prusnak** [comment]
> that it might be cleaner to just create new functions for BIP340

OK

File: crypto/schnorr.c
Comment (andrewkozlik) [inline crypto/schnorr.c:233]: Use `scalar_multiply()`, it's specially written to be faster.
```suggestion
  scalar_multiply(curve, &s, &sG);
```

**Review by andrewkozlik — CHANGES_REQUESTED** [review]

**Comment by onvej-sl** [comment]
@Fabcien, please, leave conversation resolving to reviewers.

File: crypto/schnorr.c
Comment (andrewkozlik) [inline crypto/schnorr.c:174]: Use `scalar_multiply()`, see https://github.com/trezor/trezor-firmware/pull/1509/files#r589447096.

File: crypto/schnorr.c
Comment (andrewkozlik) [inline crypto/schnorr.c:43]: Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this `bn_mod` too, but in fact this `bn_mod` seems redundant, since `n < k` is guaranteed given that `_n` is always a point coordinate and `_k` is the order of the finite field. I'd just add `bn_is_less(_n, _k)` to the `assert` above.

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:139]: `sizeof(pub_key)` is better than `33`.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:231]: > bn_subtract(&curve->order, &e, &e); seems more appropriate than bn_cnegate(), but I leave it to @onvej-sl to comment.

I totally agree.

Comment (onvej-sl) [reply crypto/schnorr.c:231]: > I am also wondering what to do if e comes out zero, in which case anyone can easily forge a signature. The likelihood of that happening by chance is infinitesimal, but it could be induced by a fault injection. I don't think we will be using Schnorr signatures in any situation where fault injections would be a concern, so I don't think we need to handle e==0, but I am asking for comments just in case anyone has a different opinion.

This is analogous to https://github.com/Fabcien/trezor-firmware/blob/add_schnorr_crypto/crypto/ecdsa.c#L1044_L1051. Is there a reason to handle it differently?

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:180]: You probably wanted `bn_mod(&R.x, &curve->prime)` but this is not necessary because the coordinates of the result of `scalar_multiply` (and `point_multiply`) are already fully reduced. (I agree it is badly documented).

Interesting is that no test failed but this is because `log2((prime - order) / prime)` is about `-128`, so it's very unlikely the coordinate was actually reduced.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:178]: This is true, since `k` goes to `bn_addmod` which cope with even unreduced numbers.

File: crypto/schnorr.c
Comment (onvej-sl) [inline crypto/schnorr.c:193]: Calling `bn_fast_most` is not necessary since an output of `bn_addmod` is partly reduced.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:43]: > Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this bn_mod too, but in fact this bn_mod seems redundant, since n < k is guaranteed given that _n is always a point coordinate and _k is the order of the finite field.

Yes, this is because the coordinates of a point that is an output of `scalar_multiply`, `point_multiply` or `point_add` are fully reduced.

> I'd just add bn_is_less(_n, _k) to the assert above.

I agree.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:62]: I thought I commented it too, but I probably didn't...

The problem is that the function `bn_fast_mod(n, k)` you want to use works only for `2^256 - 2^224 <= k <= 2^256`. So we don't have a function for general reduction.

> so you might as well replace this line with bn_subtract(&n, &k, &n),

You can't, try `jacobi(7, 1)`.

I think the algorithm works correctly. My experimental results show that it needs about three times more iterations in the outer loop for 256-bit numbers. If the quadratic reciprocity is used only if `n < k` (so it either subtracts or uses the quadratic reciprocity) the algorithm needs only two times more iterations. That's not so bad.

File: crypto/schnorr.c
Comment (andrewkozlik) [reply crypto/schnorr.c:231]: We use ECDSA to verify signatures in the bootloader, where fault injections would be a concern, but I am not sure that we will be using Schnorr signatures for anything similar. On the other hand it wouldn't hurt to fail if `e==0`, so let's err on the side of caution and do so.

File: crypto/schnorr.c
Comment (andrewkozlik) [inline crypto/schnorr.c:56]: Swapping pointers is more efficient than copying.

File: crypto/schnorr.c
Comment (andrewkozlik) [reply crypto/schnorr.c:62]: I never suggested using `bn_fast_mod()`. My point is that the `bn_mod()` here is highly misleading, because it will result in a different flow than one would expect and even when one realizes this, it's still far from obvious that the return value will be correct.

Comment (andrewkozlik) [reply crypto/schnorr.c]: Fixed in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee. @onvej-sl please review and resolve.

File: crypto/schnorr.c
Comment (andrewkozlik) [reply crypto/schnorr.c]: Done in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee.

File: crypto/schnorr.c
Comment (andrewkozlik) [reply crypto/schnorr.c]: Actually I take back my comment about adding the assert, because the algorithm works correctly even if the assertion doesn't hold, so it would be misleading to have it there. I just removed the `bn_mod()` in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee.

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c:139]: I don't think this can work since `pub_key` is a parameter, this would be `sizeof(uint8_t *)`

File: crypto/schnorr.h
Comment (Fabcien) [reply crypto/schnorr.h]: Good catch, done in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: I inlined `init_rfc6979_schnorr()` as suggested in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Removed in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Removed in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Removed in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: I added the check in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Nice ! Done in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Done in 1c7bd82

File: crypto/schnorr.c
Comment (Fabcien) [reply crypto/schnorr.c]: Removed in 1c7bd82

**Comment by Fabcien** [comment]
Thanks for the great review! I hopefully addressed all your comments in 1c7bd82. This is a fairly large change, please let me know if you want me to squash the commits.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c:123]: Absolutely right.

File: crypto/schnorr.c
Comment (onvej-sl) [reply crypto/schnorr.c]: > I never suggested using bn_fast_mod().

And I didn't mean it's you how want to use it. :)

Comment (onvej-sl) [reply crypto/schnorr.c]: Looks good!

**Review by andrewkozlik — APPROVED** [review]
> Thanks for the great review! I hopefully addressed all your comments in 1c7bd82. This is a fairly large change, please let me know if you want me to squash the commits.

Good work @Fabcien. Thank you for your contribution. I already squashed the commits in https://github.com/trezor/trezor-firmware/tree/add_schnorr_crypto and we'll see if the CI tests pass: https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/267555129. If they succeed, then I will force-push the squashed branch into this PR and we'll merge.

**Comment by andrewkozlik** [comment]
CI tests successful. Ready to merge once @onvej-sl and @prusnak approve.

**Comment by onvej-sl** [comment]
I approve (I'm not able to find the "Add review" button).

**Comment by prusnak** [comment]
Thanks! Great work everyone!

**Pull request merged**
